File: src/remote/adapter.zig
//! Backend Adapter trait.
//!
//! The type-erased `Adapter` struct mirrors `proc.Runner` (see
//! `src/proc/runner.zig`): a `*anyopaque` context and a `VTable` of function
//! pointers, with thin `pub` methods that thunk through the vtable. Concrete
//! adapters (`github`, `jira`, and the test `FakeAdapter`) plug into this
//! trait via `adapter()` helpers exactly like `RealRunner` / `FakeRunner` do
//! for `Runner`.
//!
//! The contract types this trait exchanges with callers — `MutationView`,
//! `BackendItemSnapshot`, `Outcome`, `Receipt`, `Failure`, `MutationPayload` —
//! live in `src/domain/` because they are pure data shared by `store/`,
//! `remote/`, and the future `sync/` engine. This module keeps only the trait
//! itself plus its environment-failure error sets (which are inherently tied
//! to the `proc.Runner` boundary every v1 adapter uses).

const std = @import("std");
const Allocator = std.mem.Allocator;

const Diagnostic = @import("../store/diagnostic.zig").Diagnostic;
const RunnerError = @import("../proc/runner.zig").Error;
const BackendItemSnapshot = @import("../domain/backend_item_snapshot.zig").BackendItemSnapshot;
const MutationView = @import("../domain/mutation_view.zig").MutationView;
const Outcome = @import("../domain/outcome.zig").Outcome;

/// Error set returned by `Adapter.applyMutation`.
///
/// Aliased to `proc.Runner.Error` because every v1 adapter (GitHub via `gh`,
/// Jira via `acli`) reaches the backend through the injectable subprocess
/// runner — the env-failure vocabulary is therefore exactly the runner's
/// (`ExecutableNotFound`, `SpawnFailed`, `OutOfMemory`). Mutation-level
/// rejection — non-zero exit, refused write, validation failure — flows
/// through the typed `Outcome.failure` arm so the engine can persist the
/// failure record in `mutations.failure_json` without conflating it with
/// adapter unavailability.
pub const ApplyError = RunnerError;

/// Error set returned by `Adapter.pullBackendItems`.
///
/// Extends `proc.Runner.Error` with `PullFailed` for adapter-level rejection
/// (the CLI ran but exited non-zero). `PullFailed` is paired with the
/// optional `?*Diagnostic` out-param: the adapter captures CLI stderr into
/// the diagnostic before returning the bare tag. The engine renders
/// `diag.message()` and stops sync — pull is all-or-nothing in v1 because
/// the snapshot model assumes a consistent backend view.
pub const PullError = RunnerError || error{PullFailed};

/// Type-erased Backend Adapter.
///
/// Modelled on `proc.Runner` (see `src/proc/runner.zig`): commands and the
/// sync engine consume this trait so tests can substitute a `FakeAdapter`
/// without spawning real backend CLIs.
pub const Adapter = struct {
    /// Implementation-owned pointer passed back to the vtable.
    context: *anyopaque,
    /// Adapter operations for the concrete implementation behind `context`.
    vtable: *const VTable,

    /// Adapter implementation hooks.
    pub const VTable = struct {
        pullBackendItems: *const fn (
            context: *anyopaque,
            gpa: Allocator,
            diag: ?*Diagnostic,
        ) PullError![]BackendItemSnapshot,
        applyMutation: *const fn (
            context: *anyopaque,
            gpa: Allocator,
            view: MutationView,
            now: []const u8,
        ) ApplyError!Outcome,
    };

    /// Fetch the current snapshot of backend-owned Items.
    ///
    /// On `PullError.PullFailed`, the optional `?*Diagnostic` carries CLI
    /// stderr captured by the adapter; the engine renders `diag.message()`
    /// on the next stderr line and stops the sync.
    pub fn pullBackendItems(
        self: Adapter,
        gpa: Allocator,
        diag: ?*Diagnostic,
    ) PullError![]BackendItemSnapshot {
        return self.vtable.pullBackendItems(self.context, gpa, diag);
    }

    /// Apply one pending Mutation Log entry to the backend.
    ///
    /// Returns `Outcome.success` (a `Receipt`) or `Outcome.failure` (a
    /// `Failure` whose `detail` is owned by `gpa`). Environment failures
    /// arrive as bare `ApplyError` tags via the error union.
    pub fn applyMutation(
        self: Adapter,
        gpa: Allocator,
        view: MutationView,
        now: []const u8,
    ) ApplyError!Outcome {
        return self.vtable.applyMutation(self.context, gpa, view, now);
    }
};

File: crates/tk/src/remote/adapter.rs
//! Backend Adapter trait.
//!
//! Commands and the sync engine consume this trait so tests can substitute a
//! `crate::remote::fake::FakeAdapter` without spawning real backend CLIs.
//! The contract data types it exchanges are pure domain data under
//! [`crate::domain`]. Edits and creation are separate calls because creation
//! has a stronger result-certainty contract.
//!
//! Operational calls take `&mut self` so an Adapter may consume per-call
//! state. Capability resolution also takes `&mut self` because a requested
//! repository-specific facet may require a Backend read.

use crate::domain::backend_kind::BackendKind;
use crate::domain::backend_operation::{
    AdoptedItem, BackendCreate, BackendEdit, BackendItemAddress, BackendItemInspection, BackendPull,
};
use crate::domain::backend_outcome::{BackendCreateOutcome, BackendEditOutcome};
use crate::domain::promotion_capability::{PromotionCapabilities, PromotionRequirements};
use crate::proc::ProcError;
use thiserror::Error;

/// Environment error returned by Backend Adapter edit operations.
///
/// Exactly the subprocess runner's error set: an Adapter reaches the Backend
/// through the injectable runner, so the environment-failure vocabulary is
/// the runner's. Mutation-level
/// rejection does not flow here; it rides the typed outcome so the engine can
/// persist it without conflating it with Adapter unavailability (ADR-0009).
pub type ApplyError = ProcError;

/// Error returned by Backend Adapter read operations.
///
/// Extends the runner's environment failures with [`AdapterReadError::Failed`]
/// for adapter-level rejection after the CLI ran. The payload carries the
/// Adapter's diagnostic across the typed boundary; Adopt renders it directly,
/// while Backend Pull stops before merging any collected refreshes.
#[derive(Debug, Error)]
pub enum AdapterReadError {
    /// Adapter unavailable — the Backend CLI did not start or its outcome
    /// could not be observed.
    #[error(transparent)]
    Env(#[from] ProcError),
    /// The backend CLI ran but rejected the read; the payload is the
    /// Adapter-owned diagnostic rendered by the calling command.
    #[error("{0}")]
    Failed(String),
}

/// Type-erased Backend Adapter.
pub trait Adapter {
    /// Typed kind of the Backend this Adapter reaches.
    fn backend_kind(&self) -> BackendKind;

    /// Canonicalize one Backend issue into full intake data for `tk adopt`.
    fn adopt_ticket(&mut self, input: &str) -> Result<AdoptedItem, AdapterReadError>;

    /// Refresh backend-owned fields for the exact requested working set.
    ///
    /// A successful Pull has one ordered result for every requested address
    /// and no extras. A failure prevents any refresh from being merged.
    fn pull(&mut self, items: &[BackendItemAddress]) -> Result<BackendPull, AdapterReadError>;

    /// Inspect one Backend object for Promotion recovery.
    ///
    /// This narrow read returns canonical identity, content, and Ticket Kind.
    /// It does not map Backend lifecycle state.
    fn inspect_item(&mut self, key: &str) -> Result<BackendItemInspection, AdapterReadError>;

    /// Apply one non-Promotion Mutation to an existing Backend object.
    ///
    /// Environment failures arrive through the [`ApplyError`] error arm.
    fn apply_edit(&mut self, edit: &BackendEdit) -> Result<BackendEditOutcome, ApplyError>;

    /// Apply one Promotion by creating a new Backend object.
    ///
    /// The result distinguishes a confirmed identity, certified no effect,
    /// and an indeterminate result. Creation exposes no error arm because the
    /// Adapter owns the effect-certainty classification, including whether a
    /// runner error proves the process never started.
    fn create_item(&mut self, create: &BackendCreate) -> BackendCreateOutcome;

    /// Resolve the Backend capabilities required by one Promotion graph
    /// (ADR-0036, ADR-0021).
    ///
    /// Repository-invariant facets may resolve from static data. A requested
    /// repository-specific facet may read the Backend and return an
    /// [`AdapterReadError`].
    fn resolve_promotion_capabilities(
        &mut self,
        requirements: PromotionRequirements,
    ) -> Result<PromotionCapabilities, AdapterReadError>;
}
